### DISCUSSION TITLE
Unable to authenticate from .NET

### DISCUSSION THREAD

**Discussion by sevensolutions** [zitadel/zitadel#7427] [Q&A]
I've created a very basic .NET sample.

```csharp
const string apiUrl = "https://login.myserver.com";
var key = File.ReadAllText("key.json");
var serviceAccount = ServiceAccount.LoadFromJsonString(key);

var sa = ITokenProvider.ServiceAccount(apiUrl, serviceAccount, new() { ApiAccess = true });

var client = Clients.ManagementService(new(apiUrl, sa));

var result = await client.GetMyOrgAsync(new Zitadel.Management.V1.GetMyOrgRequest());

Console.WriteLine($"User: {result}");
```

This returns:
`Unhandled exception. Grpc.Core.RpcException: Status(StatusCode="Unimplemented", Detail="Bad gRPC response. HTTP status code: 404")`

When i change the URL and add a trailing slash, i get a different error:

`Unhandled exception. Grpc.Core.RpcException: Status(StatusCode="Internal", Detail="Error starting gRPC call. HttpRequestException: {"error":"server_error","error_description":"audience is not valid: Audience must contain client_id \"https://login.myserver.com\""}`

The original sample from here is working fine: https://github.com/smartive/zitadel-net/blob/main/examples/Zitadel.ApiAccess/Program.cs

The key.json is from a service account which also has roles and projects assigned to.

Does anyone have an idea whats the problem here?

**Comment by hifabienne** [comment]
@adlerhurst @livio-a can you help here?

**Reply by livio-a** [reply]
Hey @sevensolutions 

Do you run ZITADEL self-hosted?
Because from the errors I'd guess that it is a routing problem of GRPC (first error), where you most likely got an access token.

**Reply by sevensolutions** [reply]
Yes i'am self hosting and Zitadel is running behind a Reverse Proxy which is handling TLS.
I think this brings me to the point where i should check the proxy because right now, iam only proxying http traffic which works fine for everything else.

**Reply by livio-a** [reply]
There are docs for some reverse proxies: https://zitadel.com/docs/self-hosting/manage/reverseproxy/reverse_proxy
And please also make sure to handle HTTP2 / h2c: https://zitadel.com/docs/self-hosting/manage/http2

**Reply by sevensolutions** [reply]
Many thanks for the documentation.
I've checked my config again but it turned out, that i'am using cloudflared (Tunnel) directly and the documentation says, this has an issue. Bad..
I will check if i can find a workaround, but thanks anyway 👍

**Reply by sevensolutions** [reply]
Update: Ok i think i'am almost there. What i've done:
- Updated ZITADELs configuration to use TLS everywhere.
- Enabled HTTP2 in cloudflared, as described here: https://github.com/cloudflare/cloudflared/issues/491
  - The option is called "Attempt to connect to origin using HTTP2. Origin must be configured as https."
- Web UI is still working
- Now i get a new error from my command line app:

```
Unhandled exception. Grpc.Core.RpcException: Status(StatusCode="Cancelled", Detail="No grpc-status found on response.")
   at Program.<Main>$(String[] args) in /Users/daniel/work/ZitadelApiDemo/Program.cs:line 36
   at Program.<Main>(String[] args)
```

When adding the trailing slash again, the error is still the same.
